### PR TITLE
Fix S101: Rule should not throw IndexOutOfRangeException

### LINKED ISSUE #1006: Fix S101: Rule should not throw IndexOutOfRangeException
```Severity	Code	Description	Project	File	Line	Suppression State	Detail Description
Warning	AD0001	Analyzer 'SonarAnalyzer.Rules.CSharp.ClassAndMethodName' threw an exception of type 'System.IndexOutOfRangeException' with message 'Index was outside the bounds of the array.'.	WMIHangTest		1	Active	Analyzer 'SonarAnalyzer.Rules.CSharp.ClassAndMethodName' threw the following exception:
'Exception occurred with following context:
Compilation: WMIHangTest
SyntaxTree: C:\Download\WMIHangTest\WMIHangTest\Form1.cs
SyntaxNode: Thread thread.  [MethodDeclarationSyntax]@[356..372) (16,8)-(17,0)

System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Text.StringBuilder.get_Chars(Int32 index)
   at SonarAnalyzer.Rules.CSharp.ClassAndMethodName.IsMemberNameValid(String identifierName, String& suggestion)
   at SonarAnalyzer.Rules.CSharp.ClassAndMethodName.CheckMemberName(MemberDeclarationSyntax member, SyntaxToken identifier, SyntaxNodeAnalysisContext context)
   at SonarAnalyzer.Rules.CSharp.ClassAndMethodName.<>c.<Initialize>b__10_0(SyntaxNodeAnalysisContext c)
   at SonarAnalyzer.Helpers.DiagnosticAnalyzerContextHelper.<>c__DisplayClass0_0`1.<RegisterSyntaxNodeActionInNonGenerated>b__0(SyntaxNodeAnalysisContext c)
   at SonarAnalyzer.Helpers.SonarAnalysisContext.<>c__DisplayClass16_0`1.<RegisterSyntaxNodeAction>b__0(SyntaxNodeAnalysisContext c)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__43`1.<ExecuteSyntaxNodeAction>b__43_0(ValueTuple`2 data)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
-----
'.```

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #1006

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs [hunk 1]
```diff
@@ -278,7 +279,7 @@ private static string SuggestFixedCaseName(string input, int maxUppercaseCount)
             return new string(upper.Concat(lower).ToArray());
         }
 
-        private static IEnumerable<string> SplitToParts(string name)
+        internal static IEnumerable<string> SplitToParts(string name)
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs]: Why the change to _internal_?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs [hunk 2]
```diff
@@ -137,7 +137,8 @@ private static void CheckTypeName(SyntaxNodeAnalysisContext context)
                 return;
             }
 
-            if (IsTypeComRelated(symbol.ContainingType) ||
+            if (string.IsNullOrWhiteSpace(identifier.ValueText) ||
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs]: User is complaining about `IndexOutOfRangeException` that I cannot reproduce. This seems to be the only possible case so I am handling it here.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Rules/ClassNameTest.cs [hunk 3]
```diff
@@ -56,5 +58,31 @@ public void MethodName()
...
+            new[]
+            {
+                ("foo", new [] { "foo" }),
+                ("Foo", new [] { "Foo" }),
+                ("FFF", new [] { "FFF" }),
+                ("FfF", new [] { "Ff", "F" }),
+                ("Ff9F", new [] { "Ff", "9", "F" }),
+                ("你好", new [] { "你", "好" }),
+                ("FFf", new [] { "F", "Ff" }),
+                ("", new string[0]),
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Rules/ClassNameTest.cs]: Empty string is the only case where code is not returning any data.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Rules/ClassNameTest.cs [hunk 4]
```diff
@@ -56,5 +58,31 @@ public void MethodName()
...
+                ("3%c#account", new[] { "3", "%", "c", "#", "account" }),
+            }
+            .Select(x =>
+            (
+                actual: SonarAnalyzer.Rules.CSharp.ClassAndMethodName.SplitToParts(x.Item1).ToArray(),
+                expected: x.Item2
+            ))
+            .ToList()
+            .ForEach(x => x.actual.Should().Equal(x.expected));
+        }
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Rules/ClassNameTest.cs]: The cognitive complexity of this test would be quite a bit lower if it just called a helper method _CheckNameIsSplitCorrectly_ repeatedly.
